File: leecode/code/221/221.java
class Solution {


    public int maximalSquare(char[][] matrix) {
        int len = matrix.length;
        if (len == 0) {
            return 0;
        }
        int row = matrix[0].length;


        int[][] dp = new int[len + 1][row + 1];
        int maxp = 0;
        for (int i = 1; i <= len; i++) {
            for (int j = 1; j <= row; j++) {
                if (matrix[i - 1][j - 1] == '1') {
                    dp[i][j] = Math.min(dp[i - 1][j], Math.min(dp[i][j - 1], dp[i - 1][j - 1])) + 1;
                    maxp = Math.max(dp[i][j], maxp);
                }
            }
        }
        return maxp * maxp;
    }
}

File: leecode/code/221/221.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        row = len(matrix)
        if row == 0:
            return 0
        col = len(matrix[0])
        dp = [[0]* (col + 1) for _ in range(row + 1)]
        maxp = 0
        for i in range(1, row + 1):
            for j in range(1, col + 1):
                if matrix[i - 1][j - 1] == '1':
                    dp[i][j] = min(dp[i - 1][j], dp[i - 1][j - 1], dp[i][j - 1]) + 1
                maxp = max(maxp, dp[i][j])


        return maxp * maxp